### PR TITLE
IGNITE-17575 Create a Hash Index Storage per partition

### PR DESCRIPTION
**Pull request by sashapolo**
https://issues.apache.org/jira/browse/IGNITE-17575

### CODE REVIEW

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java [hunk 1]
```diff
@@ -70,38 +85,65 @@ void setUp() {
...
+    void testCreatePartition() {
+        MvPartitionStorage absentStorage = tableStorage.getMvPartition(0);
+
+        assertThat(absentStorage, is(nullValue()));
+
+        MvPartitionStorage partitionStorage = tableStorage.getOrCreateMvPartition(0);
+
+        assertThat(partitionStorage, is(notNullValue()));
+
+        var testData = binaryRow(new TestKey(1, "1"), new TestValue(10, "10"));
```
Comment (SammyVimes) [inline modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:105]: illegal `var`

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java [hunk 2]
```diff
@@ -70,38 +85,65 @@ void setUp() {
...
+        partitionStorage.runConsistently(() -> {
+            UUID txId = UUID.randomUUID();
 
-        assertThat(tableStorage.getOrCreateMvPartition(partitionId), is(notNullValue()));
+            RowId rowId = partitionStorage.insert(testData, txId);
 
-        assertThat(tableStorage.getMvPartition(partitionId), is(notNullValue()));
+            assertThat(partitionStorage, is(sameInstance(tableStorage.getMvPartition(0))));
+
+            assertThat(unwrap(partitionStorage.read(rowId, txId)), is(equalTo(unwrap(testData))));
```
Comment (SammyVimes) [inline modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:114]: I thought we are only testing the returnal of the partition storage

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/HashIndices.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.concurrent.ConcurrentMap;
+import org.apache.ignite.internal.rocksdb.ColumnFamily;
+import org.apache.ignite.internal.storage.index.HashIndexDescriptor;
+import org.apache.ignite.internal.storage.index.HashIndexStorage;
+import org.apache.ignite.internal.storage.rocksdb.index.RocksDbHashIndexStorage;
+
+class HashIndices {
+    private final HashIndexDescriptor descriptor;
+
+    private final ConcurrentMap<Integer, HashIndexStorage> storages = new ConcurrentHashMap<>();
```
Comment (SammyVimes) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/HashIndices.java:30]: I wonder if we actually need concurrent map here. Chances are, we only create indices in one thread and access in other thread strictly after the creation.

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/HashIndices.java (hunk 3 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/HashIndices.java:30]: Do you propose to use a synchronised method instead?

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java (hunk 1 above)
Comment (sashapolo) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:105]: This code was copied from other class. Will fix

**Review by SammyVimes — APPROVED** [review]
LGTM

**Pull request merged**